WIP: Squids More Unique

jlsnow301 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
This is a fork from my previous change to glass shards, because squids slipped 100% of the time on them.

This adds a host of features to squids, including food tastes, sounds, reactions to salt (https://youtu.be/dxQmOR_QLfQ), more relevant buffs/nerfs to having tentacles and suction cups: Stronger grip, slower run speed, no slip trait, and less max melee damage.

This creates an innate ability for squids to change to a random color every 5 seconds, complete with custom, poorly drawn icon.
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->

## Why It's Good For The Game
Squid people are normally just an appearance change. This makes them a bit more flavorful, thematic. 
<!-- Please add a short description of why you think these changes would benefit the game. If you can't justify it in words, it might not be worth adding. -->

## Changelog
:cl:
add: Squids spasm and take burn damage when touching salt
add: Squids scream and take burn damage when eating salt
add: Squids do not slip on surfaces, except glass shards (briefly)
add: Squiddly melee, grab, and death sounds
add: Ability to change colors
add: Ability icon that I shittily drew
add: Turn grey on death
tweak: Fried foods now toxic to squids
tweak: Squid's grip is slightly harder to resist out of
tweak: Squid melee max damage reduced 25%
tweak: Squids take much more elemental damage
tweak: Squids run slower
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

jlsnow301 committed:
Squidlords

ike709 commented on code/game/objects/effects/decals/cleanable/food.dm:
```diff
@@ -35,6 +35,11 @@
+	if(is_species(H, /datum/species/squid))
+		while(H.loc == src.loc)
+			H.visible_message("<span class='danger'>Your tentacles shrivel as you step over the salt! It burns!</span>")
+			H.adjustFireLoss(15, TRUE)
```
You're aware this will crit squids in a matter of seconds, right?

jlsnow301 replied:
Dropped slightly.

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,10 @@
 	if(M.has_bane(BANE_SALT))
 		M.mind.disrupt_spells(-200)
+	if(is_species(H, /datum/species/squid))
+		H.visible_message("<span class='danger'>Your tongue shrivels as you taste the salt! It burns!</span>")
```
You don't actually check if it was ingested. Add `if(method == INGEST)`. I recommend adding a slightly different effect for `else if(method == TOUCH)`.

jlsnow301 replied:
Done

ike709 reviewed (changes requested):
I'm okay with the idea of this PR , but there's still code/balancing concerns to be addressed.

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,10 @@
 		return
 	if(M.has_bane(BANE_SALT))
 		M.mind.disrupt_spells(-200)
+	if(is_species(H, /datum/species/squid))
```
`H` should be `M`.

ike709 replied:
For all of these in this proc, FYI.

jlsnow301 commented:
Will readress soon

ike709 commented on code/game/objects/effects/decals/cleanable/food.dm:
```diff
@@ -35,6 +35,11 @@
 		while(L.loc == src.loc)
 			L.adjustFireLoss(2, TRUE)
 			sleep(20)
+	if(is_species(H, /datum/species/squid))
```
All of these `H`s need to be `L`s. `H` doesn't magically equal a human. Generally, you need to either use the var you have available or create a var for the human typepath and set it to the var you have available (then check it).

jlsnow301 replied:
Noted. Didn't see that define earlier.

jlsnow301 committed:
Adjustments, fixes

Courtesy of Ike for pointing these out

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 committed:
Update food_reagents.dm

qwertyquerty commented:
This seems to make squids substantially overpowered. Address that.

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,10 +319,25 @@
-		H.visible_message("<span class='danger'>Your tongue shrivels as you taste the salt! It burns!</span>")
-		H.adjustFireLoss(25, TRUE)
-		sleep(20)
+	if(is_species(M, /datum/species/squid))
```
Christ what the hell happened here? I don't even know where to begin with all of the crap you added to this proc.

ike709 commented on code/game/objects/effects/decals/cleanable/food.dm:
```diff
@@ -31,15 +31,12 @@
 		while(L.loc == src.loc)
 			L.adjustFireLoss(2, TRUE)
+			if(is_species(L, /datum/species/squid)
+				L.visible_message("<span class='danger'>Your tentacles shrivel as you step over the salt! It burns!</span>")
```
Remove the second check for squid species and just make it a more generic message, something like "The salt! It burns!". Also this should be a `to_chat()`. 

You could also maybe add a `prob()` check to scream.

jlsnow301 replied:
This prob() check to scream is nice, will find it in code shortly

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,25 @@
+			owner.log_message("threw [I] due to a Muscle Spasm", LOG_ATTACK)
+			owner.throw_item(pick(targets))
+
+/datum/reagent/consumable/sodiumchloride/reaction_mob(mob/living/M, method=INGEST, reac_volume)
```
Literally what?

jlsnow301 replied:
https://youtu.be/dxQmOR_QLfQ

ike709 replied:
I was referring to the fact that you duped the proc instead of just doing an `if(method == INGEST)` check.

jlsnow301 replied:
Thought that one of the parameters being method == touch meant this entire block was method==touch, and I would have to create a separate block for ingest. If I could just put method==ingest in it, easy

jlsnow301 committed:
WIP

jlsnow301 committed:
Update food.dm

jlsnow301 committed:
Christ I'm so going to butcher this

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,22 @@
 		return
 	if(M.has_bane(BANE_SALT))
 		M.mind.disrupt_spells(-200)
+	if(method=INGEST && is_species(M, /datum/species/squid))
```
`=` sets a var equal to something, you want `==` which checks equality.

jlsnow301 replied:
Fuck me. Beginner (pro gamer) move

ike709 commented:
The more work you put into this, the more shitcode it becomes.

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,22 @@
+	if(method=INGEST && is_species(M, /datum/species/squid))
+		to_chat(M, "<span class='danger'>Your tongue shrivels as you taste the salt! It burns!</span>")
+		M.adjustFireLoss(5, TRUE)
+		sleep(20)
```
This isn't a loop, this sleep does nothing.

jlsnow301 replied:
Interesting. Thought this was to apply sleep(time)

ike709 replied:
Yes but in this instance sleeping does nothing.

jlsnow301 replied:
> are you even testing this code as you write it? because it shouldn't even compile
Ye, sorry, knee deep in just trying to implement things. Getting an inconsistent indentation on squidpeople.dm

jlsnow301 committed:
Shitcode is bestcode

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,22 @@
+		if(LAZYLEN(targets) && I)
+			to_chat(M, "<span class='warning'>The salt causes your arm to spasm!</span>")
+			M.log_message("threw [I] due to a Muscle Spasm", LOG_ATTACK)
+			M.throw_item(pick(targets))
```
You can simplify all of this with:
`M.throw_item(get_ranged_target_turf(M, pick(GLOB.alldirs), rand(1, 3)))`
Then you can toss out the code involving picking a target.
(You'll still want to test this).

jlsnow301 replied:
Might satisfy Travis also

jlsnow301 replied:
Should I leave get_active_held_item() in here and do a check for it with if(I)?
Don't want to get a warning with no items

ike709 replied:
`throw_item()` already has the same sanity check in it.

jlsnow301 replied:
Oh does it. Well then.

jlsnow301 committed:
Travis wyd

qwertyquerty commented:
are you even testing this code as you write it? because it shouldn't even compile

jlsnow301 committed:
Sec

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

ike709 commented:
@qwertyquerty He obviously isn't, and it doesn't, but step by step I'm transforming this from terrible code into decent code.

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,20 @@
+	if(method==INGEST && is_species(M, /datum/species/squid))
+		to_chat(M, "<span class='danger'>Your tongue shrivels as you taste the salt! It burns!</span>")
+		if(prob(25))
+			user.emote("scream")
```
`M` not `user`. You gotta stop pulling vars out of thin air.

jlsnow301 replied:
fak

jlsnow301 commented:
> @qwertyquerty He obviously isn't, but step by step I'm transforming this from terrible code into decent code.

Even a broken clock is right twice a day

qwertyquerty commented:
Does broken code compile twice a day?

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,20 @@
+		if(prob(25))
+			user.emote("scream")
+		M.adjustFireLoss(5, TRUE)
+		M.Sleeping(20, 0)
```
What's with the sleeping?

jlsnow301 replied:
Another thought was hallucinate. But I felt damage would be thematic for something that makes them spasm violently

ike709 replied:
Neither sleeping nor hallucinating makes sense here.

jlsnow301 replied:
Black powder supposedly makes plasmemes hallucinate. Are you against FUN

ike709 replied:
In this case, yes.

jlsnow301 committed:
COMMITS OUT OF THIN AIR

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

ike709 commented:
Hey, it probably compiles now.

jlsnow301 committed:
Update food_reagents.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 committed:
Nice

ike709 commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -328,11 +328,9 @@
-			M.log_message("threw [I] due to a Muscle Spasm", LOG_ATTACK)
+		M.throw_item(get_ranged_target_turf(M, pick(GLOB.alldirs), rand(1, 3)))
+		to_chat(M, "<span class='warning'>The salt causes your arm to spasm!</span>")
+		M.log_message("threw [I] due to a Muscle Spasm", LOG_ATTACK)
```
`I` no longer exists as a var.

jlsnow301 replied:
Fixing now

qwertyquerty commented:
Eggs moment

jlsnow301 committed:
DIMAS MOMENT

ike709 reviewed (approved):
The code is probably as good as it's getting (at least, as good as it's getting without more time and effort than I'm willing to invest in it)

I leave balancing this in the hands of @qwertyquerty

jlsnow301 committed:
A bit of rebalancing

On further thought, this was what I was attempting to accomplish. I got to thinking strong grabber trait will mean squids can throw things everywhere on 1 click. No. This just makes it harder to break free from their grip.

jlsnow301 commented on code/modules/mob/living/living.dm:
```diff
@@ -752,6 +752,8 @@
 		var/altered_grab_state = pulledby.grab_state
 		if(resting && pulledby.grab_state < GRAB_KILL) //If resting, resisting out of a grab is equivalent to 1 grab state higher. wont make the grab state exceed the normal max, however
 			altered_grab_state++
+		else if(is_species(pulledby, /datum/species/squid) && pulledby.grab_state < GRAB_KILL) // If the puller is a squid, suction cups make it harder to break free
```
I could probably merge this into a long IF statement, but prefer the readability

ike709 commented on code/modules/mob/living/living.dm:
```diff
@@ -752,6 +752,8 @@
 		var/altered_grab_state = pulledby.grab_state
 		if(resting && pulledby.grab_state < GRAB_KILL) //If resting, resisting out of a grab is equivalent to 1 grab state higher. wont make the grab state exceed the normal max, however
 			altered_grab_state++
+		else if(is_species(pulledby, /datum/species/squid) && pulledby.grab_state < GRAB_KILL) // If the puller is a squid, suction cups make it harder to break free
```
🤔

jlsnow301 replied:
Removed strong grabber for balance. This is a more tolerable version, just harder to break free.

qwertyquerty commented:
It still seems like you're just trying to make squids better humans. They have two massive buffs and almost no nerfs

zeskorion commented:
especially with the gay glass shard nerf.

jlsnow301 commented:
I will address both of your concerns soon. Though, the strong grabber trait is gone, I wouldn't say it's MASSIVE any more if it's just harder to break free from (doesn't actually give an agg grab on first click).

jlsnow301 committed:
Prob not working yet

ike709 commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -32,6 +46,25 @@
+
 /datum/species/squid/after_equip_job(datum/job/J, mob/living/carbon/human/H)
-	H.grant_language(/datum/language/rlyehian)	
+	H.grant_language(/datum/language/rlyehian)
```
You need to add the following to `after_job_equip()` for your action to work:
```
var/datum/action/innate/squid_change/S = new
S.Grant(H)
```

ike709 commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -2,13 +2,27 @@
+    toxic_food = FRIED
+
+    var/current_color
+    var/datum/action/innate/squid_change/color
```
This action var does nothing. Remove it. See my other comment.

ike709 commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -32,6 +46,25 @@
+	icon_icon = 'icons/mob/animal.dmi'
+	button_icon_state = "squid"
+
+/datum/action/innate/squid_change/Activate()
```
I'm not sure if actions have built-in spam prevention, so test if it does or not and I'll show you how to add it if it doesn't.

ike709 replied:
Either way it may be worth putting a 5 second delay between uses, tbh.

jlsnow301 replied:
I agree, I tried briefly (before getting burned out) finding the code but couldn't find any type of cooldown for innate abilities like this

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Trying to pass diff

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Testing

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Preconfig Color Codes

jlsnow301 committed:
Didn't write state names in icon creator.

qwertyquerty commented:
Check travis logs and fix linting issues

jlsnow301 committed:
Updated

jlsnow301 commented:
Still having issues with the cooldown on an innate ability. Should pass tests now.

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Update squidpeople.dm

ike709 commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -32,6 +43,71 @@
+	var/datum/action/innate/squid_change/S = new
+	S.Grant(H)
+
+/datum/species/squid/on_species_loss(mob/living/carbon/human/H)
```
Put this in here too:
```
var/datum/action/innate/squid_change/S = locate(/datum/action/innate/squid_change) in M.actions
S?.Remove(M)
```

jlsnow301 replied:
Done

jlsnow301 replied:
What is M here? Getting an undefined var

jlsnow301 replied:
I can't find this, reverting for now

ike709 replied:
My bad. `H` not `M`. You should be able to figure this out by now though. And this still needs to be done.

jlsnow301 committed:
Ike's changes implemented

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 committed:
Update squidpeople.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 committed:
Update squidpeople.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 committed:
Update squidpeople.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 committed:
Not color changing

jlsnow301 committed:
Changing colors... Only once, though

ike709 commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -19,19 +30,88 @@
+
+/datum/action/innate/squid_change/Activate()
+	var/mob/living/carbon/human/H = owner
+	switch(rand(1,20))
```
I wonder if there's a GLOB somewhere with a list of colors.

ike709 replied:
Having a huge switch statement for semi-random colors is kind of gross.

jlsnow301 replied:
Yeah it's getting pretty ridic. This code block was pulled from xenobio/crossbreeding/__corecross.dm. It's still only giving me white, however.

jlsnow301 replied:
Fixed

ike709 commented:
This PR is bloated enough. No new features, only code improvements and balancing. Save everything else for a future PR.

ike709 commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -19,19 +30,88 @@
+/datum/action/innate/squid_change/IsAvailable()
+    if(cooldown > world.time)
+        return FALSE
+    return TRUE
```
```suggestion
    return ..()
```

Pretty sure you need this so it runs the parent's sanity checks (assuming there are any, too tired to look).

jlsnow301 committed:
Update code/modules/mob/living/carbon/human/species_types/squidpeople.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

jlsnow301 commented:
Agreed. Once color change is done, I'm finished with it.

jlsnow301 committed:
Update squidpeople.dm

ike7

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

qwertyquerty commented:
If your changelog is up to date, this seems fairly balanced to me now.

jlsnow301 commented:
It is, just having trouble with the change color ability working: It's just giving me white. Not sure if the rand(1,20) is not working or if it's just reverting to #FFFFFF.

jlsnow301 committed:
Update beestation.dme

qwertyquerty commented:
why did you remove metacoin again

jlsnow301 committed:
Revert "Changing colors... Only once, though"

This reverts commit 4ecf70c59e3440c66abd440960349aa62faae5ba.

jlsnow301 commented:
Snuck into commits earlier. Didn't mean to include it. Didn't know how to just discard the file change, tbh

qwertyquerty commented:
Put it back

jlsnow301 committed:
Update beestation.dme

jlsnow301 committed:
Merge branch 'master' into SquidPeople-Changes

jlsnow301 committed:
Update beestation.dme

jlsnow301 committed:
Merge branch 'SquidPeople-Changes' of https://github.com/jlsnow301/BeeStation-Hornet into SquidPeople-Changes

jlsnow301 committed:
Test

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Colors are working, I'm retarded

jlsnow301 committed:
Bad colors delet dis

jlsnow301 committed:
Update squidpeople.dm

jlsnow301 committed:
Trying to resolve DIFF

jlsnow301 commented:
Need last touches on it:
a) Icon update not working, have to click or hide/show it to update icon
b) Diffs is not happy

ike709 commented:
Diff bot is broken, ignore it.

Wilchenx commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -2,13 +2,24 @@
     no_equip = list(SLOT_SHOES)
     skinned_type = /obj/item/stack/sheet/animalhide/human
-    disliked_food = FRIED
+    toxic_food = FRIED
```
this is a spongebob refrence, there is really no reason to change it to toxic

qwertyquerty commented:
Ignore icondiffbot. Is this ready?

jlsnow301 commented:
It's ready. Needs a look eventually to resolve the icon update issue. But everything else okay.

qwertyquerty commented on code/modules/mob/living/carbon/human/species_types/squidpeople.dm:
```diff
@@ -19,19 +30,48 @@
+/datum/action/innate/squid_change/Activate()
+	var/mob/living/carbon/human/H = owner
+	H.dna.species.fixed_mut_color = pick("FFA500", "B19CD9", "ADD8E6", "551A8B", "0000FF",
+    "32CD32", "D3D3D3", "2956B2", "FF0000", "00FF00", "FF69B4", "FFD700", "F58B57", "5AC18E",
```
These could probably be in a static list for the species type.

jlsnow301 replied:
Throws an undefined var if I move the elsewhere, why even use one, I thought

qwertyquerty commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,18 @@
 		return
 	if(M.has_bane(BANE_SALT))
 		M.mind.disrupt_spells(-200)
+	if(method==INGEST && is_species(M, /datum/species/squid))
```
Please pad == with spaces

qwertyquerty commented on code/modules/reagents/chemistry/reagents/food_reagents.dm:
```diff
@@ -319,6 +319,18 @@
+		if(prob(25))
+			M.emote("scream")
+		M.adjustFireLoss(5, TRUE)
+	else if(method==TOUCH && is_species(M, /datum/species/squid))
```
Ditto

jlsnow301 replied:
done

qwertyquerty replied:
Whatever you did you didn't push

qwertyquerty commented:
I'll wait for the @ike709 seal of approval before putting this to vote, but this all looks good to me balance and codewise.

jlsnow301 committed:
Update food_reagents.dm

jlsnow301 committed:
Update squidpeople.dm

Co-Authored-By: Isaac <qwertytrogi@gmail.com>

qwertyquerty commented:
I'm putting this up to vote but I still await ikes seal for merging

ike709 reviewed (changes requested):
Do the thing I said to do for the action to get removed on species loss, then this is ready.

jlsnow301 committed:
Update squidpeople.dm

Co-Authored-By: ike709 <ike709@users.noreply.github.com>

qwertyquerty commented:
So is this ready @ike709 ?

ike709 commented:
Assuming it's been tested, yes.

jlsnow301 commented:
Hosted privately. Feels nice, icon just a little wonky but manageable (all icons seem this way tbf)
